javascript/ql/src/semmle/javascript/Closure.qll

```diff
@@ -37,39 +37,162 @@ class GoogFunctionCallStmt extends ExprStmt {
   GoogProvide() { getFunctionName() = "provide" }
 
   /** Gets the identifier of the namespace created by this call. */
-  string getNamespaceId() { result = getArgument(0).(ConstantString).getStringValue() }
+  override string getNamespaceId() { result = getArgument(0).(ConstantString).getStringValue() }
 }
 
 /**
  * A call to `goog.require`.
  */
-class GoogRequire extends GoogFunctionCallStmt {
+class GoogRequire extends GoogFunctionCall, GoogNamespaceRef {
   GoogRequire() { getFunctionName() = "require" }
 
   /** Gets the identifier of the namespace imported by this call. */
-  string getNamespaceId() { result = getArgument(0).(ConstantString).getStringValue() }
+  override string getNamespaceId() { result = getArgument(0).(ConstantString).getStringValue() }
+}
+
+private class GoogRequireImport extends GoogRequire, Import {
+  /** Gets the module in which this import appears. */
+  override Module getEnclosingModule() { result = getTopLevel() }
+
+  /** Gets the (unresolved) path that this import refers to. */
+  override PathExpr getImportedPath() {
+    result = getArgument(0)
+  }
 }
 
 /**
- * A Closure module, that is, a toplevel that contains a call to `goog.provide` or
- * `goog.require`.
+ * A call to `goog.module` or `goog.declareModuleId`.
  */
-class ClosureModule extends TopLevel {
+class GoogModuleDeclaration extends GoogFunctionCallStmt, GoogNamespaceRef {
+  GoogModuleDeclaration() {
+    getFunctionName() = "module" or
+    getFunctionName() = "declareModuleId"
+  }
+
+  /** Gets the identifier of the namespace imported by this call. */
+  override string getNamespaceId() { result = getArgument(0).(ConstantString).getStringValue() }
+}
+
+/**
+ * A module using the Closure module system, declared using `goog.module()` or `goog.declareModuleId()`.
+ */
+class ClosureModule extends Module {
   ClosureModule() {
+    getAChildStmt() instanceof GoogModuleDeclaration
+  }
+
+  /**
+   * Gets the call to `goog.module()` or `goog.declareModuleId` in this module.
+   */
+  GoogModuleDeclaration getModuleDeclaration() {
+    result = getAChildStmt()
+  }
+
+  /**
+   * Gets the namespace of this module.
+   */
+  string getNamespaceId() { result = getModuleDeclaration().getNamespaceId() }
+
+  override Module getAnImportedModule() {
+    exists (GoogRequireImport imprt | 
+      imprt.getEnclosingModule() = this and
+      result.(ClosureModule).getNamespaceId() = imprt.getNamespaceId()
+    )
+  }
+
+  /**
+   * Gets the top-level `exports` variable in this module, if this module is defined by
+   * a `good.module` call.
+   *
+   * This variable denotes the object exported from this module.
+   *
+   * Has no result for ES6 modules using `goog.declareModuleId`.
+   */
+  Variable getExportsVariable() {
+    getModuleDeclaration().getFunctionName() = "module" and
+    result.getScope() = this.getScope() and
+    result.getName() = "exports"
+  }
+
+  override predicate exports(string name, ASTNode export) {
```

Comment:
Could we use `DataFlow::PropWrite` to simplify this?

Reply:
I'm not sure I follow. You are looking for assignments of the form `exports.foo = bar` or `exports = { foo: bar }`, which is exactly what `PropWrite` is designed for. I agree that using `SourceNode` would be a bad idea, but these days `PropWrite` is not directly tied to `SourceNode` any more, so I would have thought that this should be safe?

Reply:
Ok, I've rewritten it to use `PropWrite`. I'm not sure it's simpler, but I'm happy not to depend on `VarDef` anymore so let's go with that.